File: code/code322.py
from typing import List
import sys
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        arr = [0] * (amount + 1)
        for i in range(1,amount+1):
            minNum = sys.maxsize
            for _,coin in enumerate(coins):
                if i - coin >= 0:
                    minNum = min(minNum,arr[i-coin])
            arr[i] = minNum + 1
        if arr[amount] == sys.maxsize + 1:
            return -1
        return arr[amount]
if __name__ == "__main__":
    solution = Solution()
    coins = [1, 2, 5]
    amount = 11
    print(solution.coinChange(coins,amount))

File: code/code322.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        vector<int> dp(amount+1,-1);
        dp[0] = 0;
        for (auto i = 1;i <= amount;i++) {
            for (auto j = 0;j < n;j++) {
                if (i - coins[j] >= 0&&dp[i - coins[j]] != -1) {
                    if (dp[i] != -1) {
                        dp[i] = min(dp[i],dp[i - coins[j]] + 1);
                    }else {
                        dp[i] = dp[i - coins[j]] + 1;
                    }
                }
            }
        }
        return dp[amount];
    }
};
int main() {
    Solution solution;
    vector<int> coins {2};
    int ans = solution.coinChange(coins,3);
    cout << ans << endl;
}